### PR TITLE
Make limit an intermediate query clause

### LINKED ISSUE #25585: NPE when using let variable in limit clause
**Description:**
The following code should explain the issue

**Steps to reproduce:**
```ballerina
function testQueryExprWithLimitForStringResultV2() returns string {
    Person p1 = {firstName: "Alex", lastName: "George", age: 23};
    Person p2 = {firstName: "Ranjan", lastName: "Fonseka", age: 30};
    Person p3 = {firstName: "John", lastName: "David", age: 33};

    Person[] personList = [p1, p2, p3];

    string outputNameString =
                from var person in personList
                let int limitValue = 2
                where person.age >= 30
                select person.firstName+" "
                limit limitValue;

    return outputNameString;
}
```

Above code produces below error,

```
 java.lang.NullPointerException
        at org.wso2.ballerinalang.compiler.bir.optimizer.BIROptimizer$RHSTempVarOptimizer.visit(BIROptimizer.java:158)
        at org.wso2.ballerinalang.compiler.bir.model.BIRNode$BIRBasicBlock.accept(BIRNode.java:387)
        at org.wso2.ballerinalang.compiler.bir.optimizer.BIROptimizer$RHSTempVarOptimizer.lambda$visit$4(BIROptimizer.java:120)
        at java.util.ArrayList.forEach(ArrayList.java:1257)
        at org.wso2.ballerinalang.compiler.bir.optimizer.BIROptimizer$RHSTempVarOptimizer.visit(BIROptimizer.java:120)
        at org.wso2.ballerinalang.compiler.bir.model.BIRNode$BIRFunction.accept(BIRNode.java:346)
        at org.wso2.ballerinalang.compiler.bir.optimizer.BIROptimizer$RHSTempVarOptimizer.lambda$visit$1(BIROptimizer.java:97)
        at java.util.ArrayList.forEach(ArrayList.java:1257)
        at org.wso2.ballerinalang.compiler.bir.optimizer.BIROptimizer$RHSTempVarOptimizer.visit(BIROptimizer.java:97)
        at org.wso2.ballerinalang.compiler.bir.model.BIRNode$BIRPackage.accept(BIRNode.java:83)
        at org.wso2.ballerinalang.compiler.bir.optimizer.BIROptimizer.optimizePackage(BIROptimizer.java:76)
        at org.wso2.ballerinalang.compiler.bir.BIRGen.genBIR(BIRGen.java:249)
        at org.wso2.ballerinalang.compiler.CompilerDriver.birGen(CompilerDriver.java:354)
        at org.wso2.ballerinalang.compiler.CompilerDriver.compile(CompilerDriver.java:301)
        at org.wso2.ballerinalang.compiler.CompilerDriver.compilePackageSymbol(CompilerDriver.java:248)
        at org.wso2.ballerinalang.compiler.CompilerDriver.compilePackage(CompilerDriver.java:140)
        at org.wso2.ballerinalang.compiler.Compiler.compilePackages(Compiler.java:166)
        at org.wso2.ballerinalang.compiler.Compiler.compilePackage(Compiler.java:174)
        at org.wso2.ballerinalang.compiler.Compiler.compile(Compiler.java:91)
        at org.ballerinalang.test.util.BCompileUtil.compile(BCompileUtil.java:516)
        at org.ballerinalang.test.util.BCompileUtil.compile(BCompileUtil.java:503)
        at org.ballerinalang.test.util.BCompileUtil.compileOnJBallerina(BCompileUtil.java:729)
        at org.ballerinalang.test.util.BCompileUtil.compileOnJBallerina(BCompileUtil.java:624)
        at org.ballerinalang.test.util.BCompileUtil.compileOnJBallerina(BCompileUtil.java:612)
        at org.ballerinalang.test.util.BCompileUtil.compileOnJBallerina(BCompileUtil.java:605)
        at org.ballerinalang.test.util.BCompileUtil.compileOnJBallerina(BCompileUtil.java:690)
        at org.ballerinalang.test.util.BCompileUtil.compile(BCompileUtil.java:110)
        at org.ballerinalang.test.query.StringQueryExpressionTest.setup(StringQueryExpressionTest.java:40)
```

### LINKED ISSUE #25679: Make limit an intermediate query clause
**Description:**
According to the spec; limit should be an intermediate query clause. However, with the current implementation, it comes after the select or query action. Therefore $title.

```
query-pipeline := from-clause intermediate-clause*
intermediate-clause :=
   from-clause
   | where-clause
   | let-clause
   | join-clause
   | limit-clause
```

**Steps to reproduce:**
n/a

**Affected Versions:**
n/a

**OS, DB, other environment details and versions:**
n/a

**Related Issues (optional):**
https://github.com/ballerina-platform/ballerina-lang/issues/25585

### PR DESCRIPTION
**Pull request by grainier**
## Purpose
$title.

Fixes #25585
Fixes #25679

## Approach
n/a

## Samples
n/a

## Remarks
- https://github.com/ballerina-platform/ballerina-lang/issues/25682

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

**Review by mohanvive — APPROVED** [review]

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java [hunk 1]
```diff
@@ -9066,11 +9065,12 @@ private STNode parseIntermediateClause(boolean isRhsExpr) {
             case ASCENDING_KEYWORD:
             case DESCENDING_KEYWORD:
                 return parseOrderByClause(isRhsExpr);
+            case LIMIT_KEYWORD:
+                return parseLimitClause(isRhsExpr);
```
Comment (lochana-chathura) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:9069]: shall we update the grammar rule in the doc comment above this method with limit clause?

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java [hunk 2]
```diff
@@ -9034,9 +9034,8 @@ private STNode parseQueryExprRhs(STNode queryConstructType, boolean isRhsExpr) {
         STNode intermediateClauses = STNodeFactory.createNodeList(clauses);
         STNode queryPipeline = STNodeFactory.createQueryPipelineNode(fromClause, intermediateClauses);
         STNode onConflictClause = parseOnConflictClause(isRhsExpr);
-        STNode limitClause = parseLimitClause(isRhsExpr);
         return STNodeFactory.createQueryExpressionNode(queryConstructType, queryPipeline, selectClause,
-                onConflictClause, limitClause);
+                onConflictClause);
```
Comment (lochana-chathura) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:9038]: doc above this method also

**Review by lochana-chathura — CHANGES_REQUESTED** [review]
New parser changes are OK! 
Have to adjust the error handler for this.
1. End `ON_CONFLICT_CLAUSE` context after parsing the `conflict` keyword in both parser and error handler
2. Remove `LIMIT_CLAUSE`  starts in seekMatchInExpressionRhs
2. Add `LIMIT_CLAUSE` to `INTERMEDIATE_CLAUSE_START` array in error handler

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java (hunk 2 above)
Comment (grainier) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:9038]: fixed in : f2064f7

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java (hunk 1 above)
Comment (grainier) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:9069]: fixed in : f2064f7

**Review by lochana-chathura — APPROVED** [review]

**Review by mohanvive — APPROVED** [review]

**Pull request merged**